### PR TITLE
fix(batch): Fix batch task leak

### PR DESCRIPTION
**Pull request by liurenjie1024**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

As title.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features).
- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### CODE REVIEW

File: src/utils/runtime/src/lib.rs [hunk 1]
```diff
@@ -48,12 +48,12 @@ fn configure_risingwave_targets_fmt(targets: filter::Targets) -> filter::Targets
     targets
         // enable trace for most modules
         .with_target("risingwave_stream", Level::DEBUG)
-        .with_target("risingwave_batch", Level::DEBUG)
+        .with_target("risingwave_batch", Level::TRACE)
```
Comment (BugenZhao) [inline src/utils/runtime/src/lib.rs:51]: Should we revert this? 👀

File: src/frontend/src/scheduler/distributed/stage.rs [hunk 2]
```diff
@@ -445,19 +456,25 @@ impl StageRunner {
...
+                         }
+                         else {
                             // After processing all stream status, we must have sent signal (Either Scheduled or
                             // Failed) to Query Runner. If this is not true, query runner will stuck cuz it do not receive any signals.
                             if !sent_signal_to_next {
                                 // For now, this kind of situation may come from recovery test: CN may get killed before reporting status, so sent signal flag is not set yet.
                                 // In this case, batch query is expected to fail. Client in simulation test should retry this query (w/o kill nodes).
                                 return Err(TaskExecutionError("compute node lose connection before response".to_string()));
                             }
-                            break;
```
Comment (chenzl25) [inline src/frontend/src/scheduler/distributed/stage.rs:456]: Is it the root cause?

File: src/frontend/src/scheduler/distributed/stage.rs [hunk 3]
```diff
@@ -701,23 +724,21 @@ impl StageRunner {
...
-                    })
-                    .await
-                {
-                    error!(
-                        "Abort task failed, task_id: {}, stage_id: {}, query_id: {}, reason: {}",
-                        task_id, stage_id, query_id, e
-                    );
-                };
-            });
+            if let Err(e) = client
```
Comment (fuyufjh) [inline src/frontend/src/scheduler/distributed/stage.rs:763]: `spawn` is removed but the comments remain unchanged:

```
2. Send RPC to each compute node for each task asynchronously.
```

Comment (fuyufjh) [reply src/frontend/src/scheduler/distributed/stage.rs:763]: And Why? 👀

File: src/frontend/src/scheduler/distributed/stage.rs [hunk 4]
```diff
@@ -513,7 +553,7 @@ impl StageRunner {
         }
 
         if let Some(err) = terminated_chunk_stream.take_result() {
-            let stage_message = err.expect("Sender should always exist!");
+            let stage_message = err.expect("Sender should always exists!");
```
Comment (fuyufjh) [inline src/frontend/src/scheduler/distributed/stage.rs:556]: Wrong syntax 🤣

File: src/frontend/src/scheduler/distributed/stage.rs [hunk 5]
```diff
@@ -672,6 +718,19 @@ impl StageRunner {
             .await;
     }
 
+    async fn notify_stage_state_changed<F>(&self, new_state: F, msg: QueryMessage)
```
Comment (fuyufjh) [inline src/frontend/src/scheduler/distributed/stage.rs:721]: How about simply

```suggestion
    async fn notify_stage_state_changed<F>(&self, old_state: StageState, new_state: StageState, msg: QueryMessage)
```

**Review by fuyufjh — COMMENTED** [review]
LGTM for rest

File: src/frontend/src/scheduler/distributed/stage.rs (hunk 5 above)
Comment (liurenjie1024) [reply src/frontend/src/scheduler/distributed/stage.rs:721]: No, this way we can't do the check in passed closure.

File: src/frontend/src/scheduler/distributed/stage.rs (hunk 3 above)
Comment (liurenjie1024) [reply src/frontend/src/scheduler/distributed/stage.rs:763]: There is not much reason, and it was removed when debugging an issue.

File: src/frontend/src/scheduler/distributed/stage.rs (hunk 5 above)
Comment (fuyufjh) [reply src/frontend/src/scheduler/distributed/stage.rs:721]: I mean

```rust
async fn notify_stage_state_changed<F>(
    &self,
    old_state: StageState,
    new_state: StageState,
    msg: QueryMessage,
) where
    F: FnOnce(StageState) -> StageState,
{
    let mut s = self.state.write().await;
    let r = mem::replace(&mut *s, new_state);
    assert_matches!(r, old_state);

    self.send_event(msg).await;
}
```

and

```rust
self.notify_stage_state_changed(
    StageState::Started,
    StageState::Running,
    QueryMessage::Stage(StageEvent::Scheduled(self.stage.id)),
)
```

What's wrong with this?

Comment (liurenjie1024) [reply src/frontend/src/scheduler/distributed/stage.rs:721]: This signature is more limited than current approach, what if the expected old state is more than 1?

Comment (liurenjie1024) [reply src/frontend/src/scheduler/distributed/stage.rs:692]: I think this is a small style problem and this is blocking the memory control policy for yuanxin and other fixes, how about we merge it first and do it later if it's proven to be necessary? cc @fuyufjh

Comment (fuyufjh) [reply src/frontend/src/scheduler/distributed/stage.rs:692]: I was not convinced but you didn't give me a choice. 🙁

**Review by fuyufjh — APPROVED** [review]

**Pull request merged**